compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs

```diff
@@ -302,18 +302,45 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     ) -> ty::InstantiatedPredicates<'tcx> {
         let bounds = self.tcx.predicates_of(def_id);
         let result = bounds.instantiate(self.tcx, args);
-        let result = self.normalize(span, result);
+        let result = self.normalize(span, None, result);
         debug!("instantiate_bounds(bounds={:?}, args={:?}) = {:?}", bounds, args, result);
         result
     }
 
-    pub(in super::super) fn normalize<T>(&self, span: Span, value: T) -> T
+    pub(in super::super) fn normalize<T>(
+        &self,
+        span: Span,
+        hir_id: Option<hir::HirId>,
+        value: T,
+    ) -> T
     where
         T: TypeFoldable<TyCtxt<'tcx>>,
     {
-        self.register_infer_ok_obligations(
-            self.at(&self.misc(span), self.param_env).normalize(value),
-        )
+        let mut infer_ok = self.at(&self.misc(span), self.param_env).normalize(value);
+        if let Some(hir_id) = hir_id {
+            // If we have access to the HIR node that introduced an Item or Binding obligation,
+            // we expand the obligation with that information for more accurate diagnostics.
+            for (idx, o) in infer_ok.obligations.iter_mut().enumerate() {
+                match o.cause.code() {
+                    ObligationCauseCode::ItemObligation(def_id) => {
+                        o.cause = traits::ObligationCause::new(
+                            o.cause.span,
+                            o.cause.body_id,
+                            ObligationCauseCode::ExprItemObligation(*def_id, hir_id, idx),
```

Comment:
This index is being used incorrectly, and will cause bugs. That index parameter should be the Nth entry in the def-id's instantiated `predicates_of` list.

Reply:
Yes, you're right. Looking to see at how to get the right index now.

Reply:
This is probably going to be pretty difficult. You should look at the other places where `*BindingObligation`s are constructed -- they all need to be enumerating in lockstep with an `InstantiatedPredicates` to make sure the indexing is not messed up.